packages/http-client-java/generator/http-client-generator-core/src/main/java/com/microsoft/typespec/http/client/generator/core/template/ConvenienceAsyncMethodTemplate.java

```diff
@@ -153,7 +162,29 @@ private String expressionMapFromBinaryData(IType responseBodyType, IType rawType
         // TODO (weidxu): support XML etc.
         switch (mimeType) {
             case TEXT:
-                mapExpression = "protocolMethodData -> protocolMethodData.toString()";
+                if (!rawType.isNullable()) {
+                    // Dealing with a primitive type that needs to be converted.
+                    PrimitiveType primitiveType = (PrimitiveType) rawType;
+                    if (rawType == BOOLEAN) {
+                        mapExpression = "protocolMethodData -> Boolean.parseBoolean(protocolMethodData.toString())";
+                    } else if (rawType == BYTE) {
+                        mapExpression = "protocolMethodData -> Byte.parseByte(protocolMethodData.toString())";
+                    } else if (rawType == INT) {
+                        mapExpression = "protocolMethodData -> Integer.parseInt(protocolMethodData.toString())";
+                    } else if (rawType == LONG) {
+                        mapExpression = "protocolMethodData -> Long.parseLong(protocolMethodData.toString())";
+                    } else if (rawType == FLOAT) {
+                        mapExpression = "protocolMethodData -> Float.parseFloat(protocolMethodData.toString())";
+                    } else if (rawType == DOUBLE) {
+                        mapExpression = "protocolMethodData -> Double.parseDouble(protocolMethodData.toString())";
+                    } else if (rawType == CHAR) {
+                        mapExpression = "protocolMethodData -> protocolMethodData.toString().charAt(0)";
+                    } else {
+                        throw new IllegalStateException("Unexpected primitive type " + primitiveType);
+                    }
+                } else {
+                    mapExpression = "protocolMethodData -> protocolMethodData.toString()";
+                }
```

Comment:
Can we extract it to a function in base class
https://github.com/microsoft/typespec/blob/cfaed553918014dc69b19fe9c7dbf64d3891b688/packages/http-client-java/generator/http-client-generator-core/src/main/java/com/microsoft/typespec/http/client/generator/core/template/ConvenienceMethodTemplateBase.java , as this block of code is duplicated in Sync/Async template?